### PR TITLE
StorageOverrideSystem for recursive item replacement, slime people spawn with nitrogen tanks

### PR DESCRIPTION
**Pull request by Krunklehorn**
## About the PR

Adds a system to recursively search through a container looking for items to replace. Intended to be paired with contextual information to determine what to do.

This system is used to conveniently allow slime people to spawn with the proper survival gear.

## Why / Balance

Slime people starting with nitrogen tanks is difficult to do in pure yaml without rearranging the code for starting gear, or just plopping one on the ground in front of them.

We already have three separate definitions for backpacks/dufflebags/satchels, each with a paragraph of yaml for 32 different presets. More variables just means more duplication.

Instead of considering hierarchies of items prior to being spawned, this PR adds a system to recursively search a container and replace it's contents.

It may have further use as more species and round start variation is added.

## Technical details

This system is designed to solve:
- "Here is a storage container. It or it's contents are prepared to be replaced, but I can't assume I always know where they'll be. The item(s) will decide what they become, I just need a system to handle the recursion and replacement."

This system should NOT be used to solve:
- "Here is a storage container. It or it's contents need replacing. I can already predict exactly where the items are, and what they should become, I just think this random system could handle it."

The former is difficult to express and maintain in YAML. The latter can easily be handled in code.
In a future PR, storage containers themselves will also be replaceable with their contents properly transferred.

#### Typical Behavior
1) `SharedStationSpawningSystem` now raises an `ApplyStorageOverrideEvent` with an instance of `StorageOverridePrototype` and an optional string for a dictionary lookup, if any items require it.

2) In `StorageOverrideSystem`, the item and all of it's contents are ~compared against a dictionary of prototype ids~ checked for the presence of a `StorageOverrideComponent` with a matching preset.

3) If the component references the required preset, the component picks a single item to spawn and replace with.

4) The replaced item is then queued for deletion.

#### Optional Behavior
- MaxReplacements: Maximum number of items to replace before halting
- SearchReplaced: Whether to continue down into storage containers after they themselves are replaced
- AllowDrop: If the new item is not going to fit, do we halt or allow it to drop on the floor?
- FromPreset: Defers the item choice to the preset, very good when a bunch of different items want to use the same pool of results

#### Component Behavior

Generic Example
```
# Preset field is required
# Choose ONE of single, random or keyed
# Single is a string, random a list and keyed is a dictionary

- type: entity
  id: JustPlainItem
  parent: SuperBoringItem
  components:
  - type: StorageOverride
    preset: FoundInFunnyBox
    single: CoolItem
    random:
        - CoolerItem
        - CoolestItem
        - NotSoCoolItem
    keyed:
        musician: VerySeriousMusicalItem
        med: VerySeriousHealingItem
        sec: VerySeriousAndSecureItem
```

Slime Nitrogen Tank Example
```
# Resources\Prototypes\Catalog\Fills\Backpacks\StarterGear\overrides.yml
- type: storageOverride                   # A recursive search through a storage container...
  id: StartingGearSlimeBackpack           # ...for starting items in a slime person's backpack.
  preset: StartingGearSlimeBackpack
  # Optional settings go here!


# Resources\Prototypes\Catalog\Fills\Items\gas_tanks.yml
- type: entity
  id: EmergencyOxygenTankFilled
  . . .
  components:
  - type: StorageOverride
    preset: StartingGear
    single: EmergencyNitrogenTankFilled
```

## Media

vlc_S1Frwrqk66

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl: Krunk
- add: Slime people now spawn with emergency nitrogen tanks.

### CODE REVIEW

**Comment by Krunklehorn** [comment]
#21544

This could easily be used to give moths some unique 1x1 food item instead of tinned meat.

File: Content.Shared/Storage/StorageOverride.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    {
+        { "EmergencyOxygenTankFilled", "EmergencyNitrogenTankFilled" },
+        { "ExtendedEmergencyOxygenTankFilled", "ExtendedEmergencyNitrogenTankFilled" },
+    };
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<MetaDataComponent, ApplyStorageOverrideEvent>(ApplyStorageOverride);
```
Comment (metalgearsloth) [inline Content.Shared/Storage/StorageOverride.cs:26]: Use broadcast events, don't subscribe to every single entity spawning.

File: Content.Shared/Storage/StorageOverride.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    // TODO: Make new prototype to compare against!
+    private readonly string _tempProtoSpecies = "SlimePerson";
+    private readonly string _tempProtoSlotName = "back";
+    private readonly Dictionary<string, string> _tempProtoPairs = new()
+    {
+        { "EmergencyOxygenTankFilled", "EmergencyNitrogenTankFilled" },
+        { "ExtendedEmergencyOxygenTankFilled", "ExtendedEmergencyNitrogenTankFilled" },
+    };
+
```
Comment (metalgearsloth) [inline Content.Shared/Storage/StorageOverride.cs:21]: Needs removing obviously

File: Content.Shared/Storage/StorageOverride.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
+using Content.Shared.Inventory;
+using Content.Shared.Preferences;
+using Robust.Shared.Containers;
+using Robust.Shared.Utility;
+
+namespace Content.Shared.Storage;
+
+public sealed partial class StorageOverrideSystem : EntitySystem
+{
+    [Dependency] private readonly IEntityManager _entityManager = default!;
```
Comment (metalgearsloth) [inline Content.Shared/Storage/StorageOverride.cs:10]: entitymanager is already a dependency on systems.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Entity prototype IDs shouldn't be for comparison checks, at best you should be using another component / prototype to group like components together.

**Comment by Krunklehorn** [comment]
Still cooking on the comparison

**Comment by Krunklehorn** [comment]
Updated code and documentation. Compares using components that reference prototypes instead of raw prototype ids alone.

The prototype carries the preset, rules and behavior.
The component references what preset to react to.
The component decides what other items the entity may be replaced with.

**Comment by UbaserB** [comment]
Could you make the musician's two instruments randomised

**Comment by Krunklehorn** [comment]
> Could you make the musician's two instruments randomised

Yes! I just tested this. Passing job data through SharedStationSpawningSystem is required, so I will add it in a future PR.

The yaml for it will look like this...

```
- type: storageOverride
  id: StartingGearMusicianBackpack
  allowDrop: true
  preset: StartingGearMusicianBackpack
  slotName: back

- type: entity
  noSpawn: true
  parent: ClothingBackpack
  id: ClothingBackpackMusicianFilled
  components:
    - type: StorageFill
      contents:
        - id: BoxSurvival
        - id: StartingGearRandomStringInstrument
        - id: StartingGearRandomWindInstrument

- type: entity
  parent: BaseHandheldInstrument
  id: StartingGearRandomStringInstrument
  components:
  - type: StorageOverride
    preset: StartingGearMusicianBackpack
    random:
      - ElectricGuitarInstrument
      - BassGuitarInstrument
      - RockGuitarInstrument
      - AcousticGuitarInstrument
      - GuitarlessFretsInstrument
      - BanjoInstrument
      - ViolinInstrument
      - ViolaInstrument
      - CelloInstrument

- type: entity
  parent: BaseHandheldInstrument
  id: StartingGearRandomWindInstrument
  components:
  - type: StorageOverride
    preset: StartingGearMusicianBackpack
    random:
      - SaxophoneInstrument
      - AccordionInstrument
      - HarmonicaInstrument
      - ClarinetInstrument
      - FluteInstrument
      - RecorderInstrument
      - PanFluteInstrument
      - OcarinaInstrument
      - BagpipeInstrument
```

**Comment by Krunklehorn** [comment]
Better solution coming soon.

**Pull request closed without merging**